spec/Data_Parallelism.tex

```diff
@@ -280,9 +284,110 @@ \section{Forall Intents}
+The shadow variable is created at the start of the task or the
+iterator and destroyed at the end of the task/iterator.
+Within the lexical scope of the forall statement or expression,
+the variable name references the shadow variable in the task
+that executed the current yield statement, or in the iterator
+if the current yield occurs in the iterator outside any task constructs.
+In contrast to regular forall intents \rsec{Forall_Intents},
+these shadow variables are unrelated to outer variables of the same name,
+if any.
+
+The syntax of a task-private variable declaration in a forall statement's
+with-clause is:
+
+\begin{syntax}
+task-private-var-decl:
+  task-private-var-kind identifier type-part[OPT] initialization-part[OPT]
+
+task-private-var-kind:
+  `const'
+  `var'
+  `ref'
+\end{syntax}
+
+The declaration of a \chpl{const} or \chpl{var} task-private variable must
+have at least one of \sntx{type-part} and \sntx{initialization-part}.
+A \chpl{ref} task-private variable must have \sntx{initialization-part}
+and cannot have \sntx{type-part}. A \chpl{ref} shadow variable
+is a reference to the \sntx{initialization-part} as calculated at
+the start of the corresponding task or the iterator.
+\chpl{ref} shadow variables are never destroyed.
+
+\begin{craychapel}
+Currently task-private variables are not available for task constructs.
+A regular variable declared at the start of the begin/cobegin/coforall
+block can be used instead.
+\end{craychapel}
+
+\begin{chapelexample}{task-private-variable.chpl}
+In the following example, the \chpl{writeln()} statement will observe
+the first shadow variable 4 times: twice each for the yields
+{\tt "before coforall"} and {\tt "after coforall"}.
+An additional shadow variable will be created and observed twice
+for each of the three \chpl{coforall} tasks.
+\begin{chapel}
+var cnt: atomic int;                     // count our shadow variables
+record R { var id = cnt.fetchAdd(1); }
+
+iter myIter() { yield ""; }              // serial iterator, unused
+
+iter myIter(param tag) where tag == iterKind.standalone {
+  for 1..2 do
+    yield "before coforall";
+  coforall 1..3 do
+    for 1..2 do
+      yield "inside coforall";
+  for 1..2 do
+    yield "after coforall";
+}
+
+forall str in myIter()
+  with (var tpv: R)                      // delcare a task-private variable
+do
+  writeln("shadow var: ", tpv.id, "  yield: ", str);
+\end{chapel}
+\begin{chapelprediff}
+\#!/usr/bin/env sh
+testname=$1
+outfile=$2
+sort $outfile > $outfile.2
+mv $outfile.2 $outfile
+\end{chapelprediff}
+\begin{chapeloutput}
+shadow var: 0  yield: after coforall
+shadow var: 0  yield: after coforall
+shadow var: 0  yield: before coforall
+shadow var: 0  yield: before coforall
+shadow var: 1  yield: inside coforall
+shadow var: 1  yield: inside coforall
+shadow var: 2  yield: inside coforall
+shadow var: 2  yield: inside coforall
+shadow var: 3  yield: inside coforall
+shadow var: 3  yield: inside coforall
+\end{chapeloutput}
```

Comment:
Should we include this output in the spec? It could be useful to have a comment matching the integer IDs to lines in the example, like:
```chpl
for 1..2 do
  yield "before coforall"; // shadow var 0
```

Reply:
The output does not look great because it needs to be sorted so the first two lines happen to be the last two lines actually printed by the program.